dags/mag_and_friends.py

```diff
@@ -0,0 +1,491 @@
+##
+##          It should also be noted that this DAG assumes nextflow and conda are installed on the cluster.
+##
+## !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+
+# Constants for file paths
+# TODO should some/ all of these be configurable?
+CLUSTER_CONFIG_PATH = "/data/MicrobiomeDB/mbio_airflow_dags/cluster_configs/pmacs.json"
+BASE_PATH = "/data/MicrobiomeDB/common/shotgun_metagenomics"
+PROVENANCE_PATH = join(BASE_PATH, "processed_studies_provenance.csv")
+# metagenomic_studies.csv should have two columns, one w the study name and the other the directory
+ALL_STUDIES_PATH = join(BASE_PATH, "metagenomics_studies.csv")
+
+MAG_VERSION = "3.0.1"
+TAXPROFILER_VERSION = "1.1.8"
+# the wget download method has a bug in version 1.12.0
+# its fixed on dev, and we should use that for now
+# switch back to a stable version once the fix is merged and released
+FETCHNGS_VERSION = "dev"
+METATDENOVO_VERSION = "1.0.1"
+
+NEXTFLOW_VERSION = "23.10.1"
+
+# TODO i like these two functions, they should go in a utils file. 
+# maybe we could use a StudyProvenanceProcessor class or something?
+def load_processed_studies(provenance_path=PROVENANCE_PATH):
+    """Load processed studies from a CSV file into a dictionary."""
+    processed_studies = {}
+    if os.path.exists(provenance_path):
+        with open(provenance_path, 'r') as file:
+            reader = csv.DictReader(file)
+            for row in reader:
+                processed_studies[row['study']] = {'timestamp': row['timestamp'], 
+                                                    'mag_revision': row['mag_revision'],
+                                                    'metatdenovo_revision': row['metatdenovo_revision'],
+                                                    'taxprofiler_revision': row['taxprofiler_revision']}
+    return processed_studies
+
+# TODO make sure this overwrites existing rows in the provenance file
+# TODO swap to using this in the dag below
+def update_provenance_file(
+    study, 
+    timestamp, 
+    provenance_path=PROVENANCE_PATH, 
+    mag_revision=MAG_VERSION,
+    metatdenovo_revision=METATDENOVO_VERSION,
+    taxprofiler_revision=TAXPROFILER_VERSION
+):
+    """Update the provenance file with the latest processing information."""
+    fieldnames = ['study', 'timestamp', 'mag_revision', 'metatdenovo_revision', 'taxprofiler_revision']
+    new_data = {
+        'study': study, 
+        'timestamp': timestamp, 
+        'mag_revision': mag_revision,
+        'metatdenovo_revision': metatdenovo_revision,
+        'taxprofiler_revision': taxprofiler_revision
+    }
+    with open(provenance_path, 'a', newline='') as csvfile:
+        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
+        if os.stat(provenance_path).st_size == 0:
+            writer.writeheader()
+        writer.writerow(new_data)
+
+def create_dag():
+    default_args = {
```

Comment:
I was running into dag.task failing due to unable to write log file or something quite frequently before. Not sure if you have ever run into similar issues. I ended up having to add "retries" and 'retry_delay' to force retry to get around. Otherwise I have to manually clear the task error and get them restart.

Reply:
i suspect this error is to do w how we have been running airflow on the 180 server more than anything specific to airflow itself or the dags were writing. i also suspect it will disappear when i do. happy to discuss more outside this pr.